### PR TITLE
[OpenCL] Fix an infinite loop in builidng AddrSpaceQualType

### PR DESCRIPTION
**Pull request by changpeng**
In building AddrSpaceQualType (https://github.com/llvm/llvm-project/pull/90048), there is a bug in removeAddrSpaceQualType() for arrays. Arrays are weird because qualifiers on the element type also count as qualifiers on the type, so getSingleStepDesugaredType() can't remove the sugar on arrays. This results in an infinite loop in removeAddrSpaceQualType. To fix the issue, we use ASTContext::getUnqualifiedArrayType instead, which strips the qualifier off the element type, then reconstruct the array type.

### CODE REVIEW

File: clang/lib/CodeGen/CGExprAgg.cpp [hunk 1]
```diff
@@ -537,8 +537,9 @@ void AggExprEmitter::EmitArrayInit(Address DestPtr, llvm::ArrayType *AType,
       elementType.isTriviallyCopyableType(CGF.getContext())) {
     CodeGen::CodeGenModule &CGM = CGF.CGM;
     ConstantEmitter Emitter(CGF);
+    Qualifiers Quals;
     QualType GVArrayQTy = CGM.getContext().getAddrSpaceQualType(
-        CGM.getContext().removeAddrSpaceQualType(ArrayQTy),
+        CGM.getContext().getUnqualifiedArrayType(ArrayQTy, Quals),
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/CGExprAgg.cpp:542]: I see two issues with this fix:

1. It leaves removeAddrSpaceQualType broken for anyone else who tries to use it.  Ideally, removeAddrSpaceQualType should do what it says.
2. This drops other qualifiers on the floor; they might be relevant in some cases.

Comment (changpeng) [reply clang/lib/CodeGen/CGExprAgg.cpp:542]: Do you mean we should actually fix removeAddrSpaceQualType? Somewhere inside removeAddrSpaceQualType, we
should use getUnqualifiedArrayType if it is an arrayType, and getSingleStepDesugaredType othereise?
I have to admit that I have no experience in this field, so I am relying on you and @svenvh to move on for a reasonable fix. Thanks.

Comment (efriedma-quic) [reply clang/lib/CodeGen/CGExprAgg.cpp:542]: Something like that, yes.

Comment (changpeng) [reply clang/lib/CodeGen/CGExprAgg.cpp]: > Something like that, yes.

Thanks. Updated. Should be closer!

File: clang/test/CodeGenOpenCLCXX/array-type-infinite-loop.clcpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 4
+//RUN: %clang_cc1 %s -emit-llvm -O1 -o - | FileCheck %s
```
Comment (arsenm) [inline clang/test/CodeGenOpenCLCXX/array-type-infinite-loop.clcpp:2]: codegen tests need an explicit target

File: clang/lib/AST/ASTContext.cpp [hunk 3]
```diff
@@ -3054,6 +3054,13 @@ QualType ASTContext::removeAddrSpaceQualType(QualType T) const {
   if (!T.hasAddressSpace())
     return T;
 
+  // For arrays, strip the qualifier off the element type, then reconstruct the
+  // array type
+  if (T.getTypePtr()->isArrayType()) {
+    Qualifiers Qualfs;
+    return getUnqualifiedArrayType(T, Qualfs);
```
Comment (efriedma-quic) [inline clang/lib/AST/ASTContext.cpp:3061]: You need to preserve the non-address-space qualifiers, the same way the code after the loop does.  Actually, I'd just recommend restructuring to use the existing code, something like:

```
if (T.getTypePtr()->isArrayType()) {
  getUnqualifiedArrayType()
} else {
  while (T.hasAddressSpace()) {
     [...]
  }
}
```

Comment (changpeng) [reply clang/lib/AST/ASTContext.cpp:3061]: Thanks.  Can I do as the following?
Note that I am passing QualifierCollector to getUnqualifiedArrayType, which has Qualifiers as the second argument.
 Also,  TypeNode = T.getTypePtr(); after I is unqualified.

 QualifierCollector Quals;
  const Type *TypeNode;
  if (T.getTypePtr()->isArrayType()) {
    T = getUnqualifiedArrayType(T, Quals);
    TypeNode = T.getTypePtr();
  } else {
  while (T.hasAddressSpace()) {
  ....

File: clang/test/CodeGenOpenCLCXX/array-type-infinite-loop.clcpp (hunk 2 above)
Comment (changpeng) [reply clang/test/CodeGenOpenCLCXX/array-type-infinite-loop.clcpp]: add "triple spir", same as other tests in the same directory

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Pull request merged**